src/backend/opengl/DepthStencilStateGL.cpp

```diff
@@ -0,0 +1,128 @@
+                case nxt::CompareFunction::GreaterEqual:
+                    return GL_GEQUAL;
+                case nxt::CompareFunction::NotEqual:
+                    return GL_NOTEQUAL;
+                case nxt::CompareFunction::Equal:
+                    return GL_EQUAL;
+                case nxt::CompareFunction::Always:
+                    return GL_ALWAYS;
+            }
+        }
+
+        GLuint OpenGLStencilOperation(nxt::StencilOperation stencilOperation) {
+            switch (stencilOperation) {
+                case nxt::StencilOperation::Keep:
+                    return GL_KEEP;
+                case nxt::StencilOperation::Zero:
+                    return GL_ZERO;
+                case nxt::StencilOperation::Replace:
+                    return GL_REPLACE;
+                case nxt::StencilOperation::Invert:
+                    return GL_INVERT;
+                case nxt::StencilOperation::IncrementClamp:
+                    return GL_INCR;
+                case nxt::StencilOperation::DecrementClamp:
+                    return GL_DECR;
+                case nxt::StencilOperation::IncrementWrap:
+                    return GL_INCR_WRAP;
+                case nxt::StencilOperation::DecrementWrap:
+                    return GL_DECR_WRAP;
+            }
+        }
+    }
+
+    DepthStencilState::DepthStencilState(Device* device, DepthStencilStateBuilder* builder)
+        : DepthStencilStateBase(builder), device(device) {
+    }
+
+    void DepthStencilState::ApplyNow(PersistentPipelineState &persistentPipelineState) const {
+        if (DepthTestEnabled()) {
+            glEnable(GL_DEPTH_TEST);
+        } else {
+            glDisable(GL_DEPTH_TEST);
+        }
+
+        auto& depthInfo = GetDepth();
+
+        if (depthInfo.depthWriteEnabled) {
+            glDepthMask(GL_TRUE);
+        } else {
+            glDepthMask(GL_FALSE);
+        }
+
+        glDepthFunc(OpenGLCompareFunction(depthInfo.compareFunction));
+
+        if (StencilTestEnabled()) {
+            glEnable(GL_STENCIL_TEST);
+        } else {
+            glDisable(GL_STENCIL_TEST);
+        }
+
+        auto& stencilInfo = GetStencil();
+
+        GLenum backCompareFunction = OpenGLCompareFunction(stencilInfo.back.compareFunction);
+        GLenum frontCompareFunction = OpenGLCompareFunction(stencilInfo.front.compareFunction);
+
+        persistentPipelineState.CacheStencilFuncsAndMask(backCompareFunction, frontCompareFunction, stencilInfo.readMask);
+
+        glStencilOpSeparate(GL_BACK,
+            OpenGLStencilOperation(stencilInfo.back.stencilFail),
+            OpenGLStencilOperation(stencilInfo.back.depthFail),
+            OpenGLStencilOperation(stencilInfo.back.depthStencilPass)
+        );
+        glStencilFuncSeparate(GL_BACK,
+            backCompareFunction,
+            persistentPipelineState.GetCachedStencilReference(),
+            stencilInfo.readMask
+        );
+        glStencilMaskSeparate(GL_BACK, stencilInfo.writeMask);
```

Comment:
nit: the masks aren't separate so you could set both the back and front masks at the same time.